### PR TITLE
Lecture 07 idiomatic python

### PR DESCRIPTION
**Pull request by Cjkjvfnby**

### CODE REVIEW

File: lecture_04_test/materials/slides/Presentation.ipynb [hunk 1]
```diff
@@ -1153,7 +1153,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.8.1"
+   "version": "3.9.0"
```
Comment (Cjkjvfnby) [inline lecture_04_test/materials/slides/Presentation.ipynb:1156]: Does it work now?

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 2]
```diff
@@ -0,0 +1,2325 @@
...
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "0 1 2 3 4 5 "
+     ]
+    }
+   ],
+   "source": [
+    "for i in [0,1,2,3,4,5]:\n",
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb:97]: Maybe add a header before code?

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 3]
```diff
@@ -0,0 +1,2325 @@
...
+     "text": [
+      "<class 'list_reverseiterator'>\n",
+      "48\n",
+      "88\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(type(reversed(colors)))\n",
+    "print(sys.getsizeof(reversed(colors)))\n",
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb]: Items in a different order here and in the examples.

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 4]
```diff
@@ -0,0 +1,2325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Looping over two (or more) collections"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb:417]: Of the same size.  You can mention zip_longest here.

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 5]
```diff
@@ -0,0 +1,2325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Multiple exit points in loops"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb:582]: Maybe mention `yield` and that it will be described later.

**Review by Cjkjvfnby — COMMENTED** [review]
Can you add tome headings for examples
*Good* *Bad* *Wrong* *Correct* 

See examples here https://www.python.org/dev/peps/pep-0008/#should-a-line-break-before-or-after-a-binary-operator

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 6]
```diff
@@ -0,0 +1,2325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Copying"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb:778]: Will you talk about defensive copy and immutability here?

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 7]
```diff
@@ -0,0 +1,2325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Sortings"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb:952]: Bad example
```
x = list(iterable)
x.sort()
```

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 8]
```diff
@@ -0,0 +1,2325 @@
...
+      "{1: 'four', 4: 'one', 3: 'three', 2: 'two'}\n"
+     ]
+    }
+   ],
+   "source": [
+    "print({key: some_dict[key] for key in sorted(some_dict, key=lambda x: some_dict[x])})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb]: It should be a code cell.

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 9]
```diff
@@ -0,0 +1,2325 @@
...
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "By value"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb]: You don't have a heading `by key and value`

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 10]
```diff
@@ -0,0 +1,2325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "subslide"
+    }
+   },
+   "source": [
+    "Use dict comprehensions instead"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb]: You usually don't use such a style, maybe make it bold or some heading.

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 11]
```diff
@@ -0,0 +1,2325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "source": [
+    "Common way"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb]: I usually do  `new_dict = dict(first)`

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 12]
```diff
@@ -0,0 +1,2325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "source": [
+    "This is job for **join**"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb]: It is a good point to ask them why?

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 13]
```diff
@@ -0,0 +1,2325 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "### Updating seuences"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb]: Sequences?

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 14]
```diff
@@ -0,0 +1,2325 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "1.47 ms ± 38.9 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)\n"
+     ]
+    }
+   ],
+   "source": [
+    "l = list(range(1_000_000))\n",
+    "%timeit l.pop(0); l.insert(0, 42)"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb:1993]: Please also add pop right examples for the list and deque.   Both operations are O(1).

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 15]
```diff
@@ -0,0 +1,2325 @@
...
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "a wasn't reached\n"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb]: Highlight a with ticks `a`

File: lecture_07_idiomatic_python/materials/slides.ipynb [hunk 16]
```diff
@@ -0,0 +1,2325 @@
...
+    "collapsed": false,
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "### Python is lazy"
```
Comment (Cjkjvfnby) [inline lecture_07_idiomatic_python/materials/slides.ipynb:2466]: This part is duplicated

**Comment by cdeler** [comment]
@RomanTsaregorodtsev 
Occasionally I cannot left any comments in the diff, so I'll put it here

After showing:
```python
var = "" or [1,2,3] or 12
tmp = "" or 0 or False
```
I'd show:
```python
var = str(some_other_val) is some_other_val is not None else "Nothing"
val2 = "Hello" if True else 1/0  # there is no exception here
```

**Pull request merged**